### PR TITLE
Implement TracingDiagnosticListener as a WorkflowInterceptor.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
Two commits, first commit keeps it implementing the old `WorkflowDiagnosticListener` and just moves the golden value test over, and then the second commit removes the `WorkflowDiagnosticListener` interface and converts call sites.

There aren't many interesting changes in the golden value text file, as you'd expect. The only change is actually that `onPropsChanged` for the root and for the root workflow are called in slightly different places, and the worker IDs are different. Everything else is exactly the same. The changes to it in the second commit just update the name.

## Checklist

- [x] Unit Tests
- [x] UI Tests
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: samples/dungeon/app/src/main/java/com/squareup/sample/dungeon/MainActivity.kt [hunk 1]
```diff
@@ -34,7 +34,7 @@ class MainActivity : AppCompatActivity() {
       WorkflowRunner.Config(
           workflow = component.timeMachineWorkflow,
           props = "simple_maze.txt",
-          diagnosticListener = TracingDiagnosticListener(traceFile)
+          interceptors = listOf(TracingWorkflowInterceptor(traceFile))
```
Comment (rjrjr) [inline samples/dungeon/app/src/main/java/com/squareup/sample/dungeon/MainActivity.kt:37]: SOOOoooo much better.

File: workflow-runtime/src/test/java/com/squareup/workflow/internal/RealRenderContextTest.kt [hunk 2]
```diff
@@ -140,6 +140,7 @@ class RealRenderContextTest {
     val handler = context.onEvent<String> { expectedUpdate }
     assertTrue(eventActionsChannel.isEmpty)
 
+    context.freeze()
```
Comment (rjrjr) [inline workflow-runtime/src/test/java/com/squareup/workflow/internal/RealRenderContextTest.kt:143]: Seems unrelated. Opportunistic cleanup?

Comment (zach-klippenstein) [reply workflow-runtime/src/test/java/com/squareup/workflow/internal/RealRenderContextTest.kt:143]: Nope, required because above change makes `onEvent` delegate to `send`, which will throw if we don't freeze the context before sending to it. I also want to make `onEvent` an extension method on the context.
